python/dgl/dataloading/dataloader.py

```diff
@@ -575,38 +620,53 @@ def _collate(self, items):
         if isinstance(items[0], tuple):
             # returns a list of pairs: group them by node types into a dict
             items = utils.group_as_dict(items)
-            items = utils.prepare_tensor_dict(self.g_sampling, items, 'items')
+            items = _prepare_tensor_dict(self.g_sampling, items, 'items', self._is_distributed)
         else:
-            items = utils.prepare_tensor(self.g_sampling, items, 'items')
+            items = _prepare_tensor(self.g_sampling, items, 'items', self._is_distributed)
 
-        pair_graph = self.g.edge_subgraph(items)
+        if isinstance(items, dict):
+            sample_items = {k: F.gather_row(self.eids[k], v) for k, v in items.items()} \
+                           if self.return_indices else items
+        else:
+            sample_items = F.gather_row(self.eids, items) if self.return_indices else items
+
+        pair_graph = self.g.edge_subgraph(sample_items)
         seed_nodes = pair_graph.ndata[NID]
 
         exclude_eids = _find_exclude_eids(
             self.g,
             self.exclude,
-            items,
+            sample_items,
             reverse_eid_map=self.reverse_eids,
             reverse_etype_map=self.reverse_etypes)
 
         blocks = self.block_sampler.sample_blocks(
             self.g_sampling, seed_nodes, exclude_eids=exclude_eids)
         input_nodes = blocks[0].srcdata[NID]
 
-        return input_nodes, pair_graph, blocks
+        if not self.return_indices:
+            return input_nodes, pair_graph, blocks
+        else:
+            return input_nodes, pair_graph, items, blocks
 
     def _collate_with_negative_sampling(self, items):
         if isinstance(items[0], tuple):
             # returns a list of pairs: group them by node types into a dict
             items = utils.group_as_dict(items)
-            items = utils.prepare_tensor_dict(self.g_sampling, items, 'items')
+            items = _prepare_tensor_dict(self.g_sampling, items, 'items', self._is_distributed)
         else:
-            items = utils.prepare_tensor(self.g_sampling, items, 'items')
+            items = _prepare_tensor(self.g_sampling, items, 'items', self._is_distributed)
 
-        pair_graph = self.g.edge_subgraph(items, preserve_nodes=True)
+        if isinstance(items, dict):
+            sample_items = {k: F.gather_row(self.eids[k], v) for k, v in items.items()} \
+                           if self.return_indices else items
```

Comment:
One question about negative sampling is that can we sample pos edges belong to a specific positive edge set and sample negative edges from another edge set?

Reply:
Can you do it by writing your own negative sampler?  Do you also need indices?

Reply:
Writing my own negative sampler always can work.